topology/probes/runc/hosts.go

```diff
@@ -0,0 +1,97 @@
+// +build linux
+
+/*
+ * Copyright (C) 2019 IBM, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy ofthe License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specificlanguage governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package runc
+
+import (
+	"bufio"
+	"net"
+	"os"
+	"strings"
+)
+
+// hosts contains raw host entries.
+type hosts struct {
+	IP       string              `json:"IP,omitempty"`
+	Hostname string              `json:"Hostname,omitempty"`
+	ByIP     map[string][]string `json:"ByIP,omitempty"`
+}
+
+func newHosts() *hosts {
+	hosts := new(hosts)
+	hosts.ByIP = make(map[string][]string)
+	return hosts
+}
+
+// Count occurrences in s of any bytes in t.
+func countAnyByte(s string, t string) int {
+	n := 0
+	for i := 0; i < len(s); i++ {
+		if strings.IndexByte(t, s[i]) >= 0 {
+			n++
+		}
+	}
+	return n
+}
+
+func readHosts(path string) (*hosts, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	hosts := newHosts()
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if err := scanner.Err(); err != nil {
+			return nil, err
+		}
+
+		if i := strings.IndexByte(line, '#'); i >= 0 {
+			// Discard comment.
+			line = line[0:i]
+		}
+
+		f := strings.Fields(line)
+		if len(f) < 2 {
+			continue
+		}
+
+		ip := net.ParseIP(f[0])
+		if ip == nil {
+			continue
+		}
+
+		if ip.IsLoopback() {
+			continue
+		}
+
+		hosts.IP = ip.String()
```

Comment:
We need to handle cases in which there is more than a single IP in the /etc/hosts (that is not loopback) ... the current solution is not guarantied to work in such cases in any consistent way.

You might take the largest or smallest IP ... or sort lexicographic .... or something ... in any event taking the 'last' IP in the table might not be consistent over reboots or other operations done by k8s